server/routerlicious/packages/routerlicious-base/src/riddler/tenantManager.ts

```diff
@@ -455,4 +558,44 @@ export class TenantManager {
         const accessInfo = JSON.parse(this.secretManager.decryptSecret(encryptedAccessInfo));
         return accessInfo;
     }
+
+    private decryptCachedKeys(cachedKey: string) {
+        const keys = JSON.parse(cachedKey);
+        return keys.key2 === "" ? {
+                key1: this.secretManager.decryptSecret(keys.key1),
+                key2: "",
+            } : {
+                key1: this.secretManager.decryptSecret(keys.key1),
+                key2: this.secretManager.decryptSecret(keys.key2),
+            };
+    }
+
+    private async getCacheKey(tenantId: string): Promise<string> {
+        return this.cache?.get(`tenantKeys:${tenantId}`);
+    }
+
+    private async deleteCacheKey(tenantId: string): Promise<boolean> {
+        return this.cache?.delete(`tenantKeys:${tenantId}`);
+    }
+
+    private async setCacheKey(tenantId: string, value: ITenantKeys, metric?: Lumber<LumberEventName.RiddlerFetchTenantKey>): Promise<void> {
+        const lumberProperties = { [BaseTelemetryProperties.tenantId]: tenantId };
+        try {
+            await this.cache?.set(`tenantKeys:${tenantId}`, JSON.stringify(value));
+            Lumberjack.info(`Added new tenant keys to cache.`, lumberProperties);
+        } catch (error) {
+            if (metric) {
+                metric.setProperty("CacheFailed", true);
+            }
+            Lumberjack.error(`Reading from cache failed`, lumberProperties, error);
+            return undefined;
+        }
+        await this.cache?.set(`tenantKeys:${tenantId}`, JSON.stringify(value)).catch((error)=>{
+            if (metric){
+                metric.setProperty("CacheFailed", true);
+            }
+            Lumberjack.error(`Error adding keys to cache.`, lumberProperties, error)
+        });
```

Comment:
Looks like this section is repeating lines 583-592. Perhaps this is leftover code that was mistakenly not removed, when switching to try/catch block?

Reply:
That is exactly what happened, thank you for catching this

Reply:
The two duplicate logics seem still exists here?

Reply:
Also, pass in `metric` works here but it seems the SOLID pattern.
Maybe from caller, you can pass in a correlation ID instead (not necessary the correlation ID from http request, because it would hard to track. maybe just create a cache corrlationId)? Set that `corrlationId` in caller metric

And just modify the code there to use a metric `SetCache` with `correlationID`, and just set the metric sucess if no exception, and failed the metric with exception?

`Correlationed` might not even needed because with your changes, I bet `setCache` would got called a lot, we can probrobaly use tenantId to correlate?

Reply:
@zhangxin511 , is your suggestion to instead have 2 metrics, the second one being a new metric to be defined here in `setCache()`, and then correlate both with a new correlationId?

For this particular use case scenario, I think it might be ok to keep it as is just for the sake of simplification. Querying and log analysis would be penalized if we have to correlate 2 metrics because of this design change. The way it currently is allows to have all the info within the same telemetry data (that is, in the same log line). We seem to be doing something similar here for ScribeSession Metric - I think @pradeepvairamani added that code

https://github.com/microsoft/FluidFramework/blob/4f92d391882f60f850cd5995a80e287643a2057b/server/routerlicious/packages/lambdas/src/scribe/lambdaFactory.ts#L224

Reply:
We can use two metrics, or caller metric + normal log. callee should have no knowledge/dependencies of caller.

Reply:
We talked offline and we decided that the easiest and cleanest option for now is to make `setKeyInCache` return a `Promise<boolean>` instead of `Promise<void>`. The boolean would represent whether the operation succeeded or not. Then, the caller could use the response boolean to set a property like `settingKeyInCacheSucceeded` in the metric it is using. That would fix the problem Xin pointed out, and at the same time, keep all information into a single metric/log line. Something like this (could be applied in line 342 of tenantManager.ts and also elsewhere):

```ts
const setKeyInCacheSucceeded = await this.setKeyInCache(tenantId, cacheKeys);
fetchTenantKeyMetric.setProperty("settingKeyInCacheSucceeded", setKeyInCacheSucceeded);
```

Reply:
Since the fetchTenantKeyMetric function is the only location we are currently adding metrics, and we have logs within the setKeyInCache functions for success/failure, I am just implementing this change at line 343.